File: Python/find_all_duplicates.py
"""
Given an integer array nums of length n where all the integers of nums are in the range [1, n] and each integer
appears once or twice, return an array of all the integers that appears twice.
"""


def solve(nums):
    # using set
    set_list = set()
    ans = []
    for e in nums:
        if e not in set_list:
            set_list.add(e)
        else:
            ans.append(e)
    return ans


def solve2(nums):
    # using cyclic sort
    ans = []
    i = 0
    while i < len(nums):
        correct = nums[i]-1
        if nums[i] != nums[correct]:
            nums[i], nums[correct] = nums[correct], nums[i]
        else:
            i += 1

    for i in range(len(nums)):
        if nums[i] != i+1:
            ans.append(nums[i])

    return ans


def main():
    nums = [4, 3, 2, 7, 8, 2, 3, 1]
    print(solve(nums))
    print(solve2(nums))


if __name__ == '__main__':
    main()

File: Java/src/FindAllDuplicates.java
/*
Given an integer array nums of length n where all the integers of nums are in the range [1, n] and
each integer appears once or twice, return an array of all the integers that appears twice.
 */


import java.util.ArrayList;
import java.util.HashSet;

public class FindAllDuplicates {
    static ArrayList<Integer> solve(int[] nums) {
        ArrayList<Integer> list = new ArrayList<>();
        HashSet<Integer> seen = new HashSet<>();

        for (int e: nums) {
            if (!seen.contains(e)) seen.add(e);
            else list.add(e);
        }
        return list;
    }

    public static void main(String[] args) {
        int[] nums = {1, 2, 3, 4, 1, 2};
        System.out.println(solve(nums));
    }
}
